### PR TITLE
Feature/se 410 derived property tests

### PR DESCRIPTION
**Pull request by cscialino**
# Pull Request Checklist

- [x ] Read the [contributing guidelines](../docs/CONTRIBUTING.md)
- [x ] Tests pass
- [ ] Raised the PR against the `develop` branch

# Description of the PR

PR for derived properties test, using coalesce for ratings data

### CODE REVIEW

File: sdk/tests/tutorials/properties/test_derived_properties.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
+import unittest
+
```
Comment (slemasne) [inline sdk/tests/tutorials/properties/test_derived_properties.py:2]: Run `black`  and clean up all this extra spacing.

File: sdk/tests/tutorials/properties/test_derived_properties.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    def create_ratings_property(self, figi, fitch_value=None, moodys_value=None):
+
+        ratings = ["Fitch", "Moodys"]
+
+        for rating in ratings:
+            property_definition = models.CreatePropertyDefinitionRequest(
+                domain="Instrument",
+                scope="Test-Demo",
+                code="{}Rating".format(rating),
```
Comment (slemasne) [inline sdk/tests/tutorials/properties/test_derived_properties.py:33]: Use f-strings. These look cleaner.

f"foo {bar}"

File: sdk/tests/tutorials/properties/test_derived_properties.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+                data_type_id=lusid.ResourceId(
+                    scope="system", code="number"))
+
+            try:
+                # create property definition
+                self.property_definitions_api.create_property_definition(
+                    create_property_definition_request=property_definition)
+            except lusid.ApiException as e:
+                if json.loads(e.body)["name"] == "PropertyAlreadyExists":
+                    pass
```
Comment (slemasne) [inline sdk/tests/tutorials/properties/test_derived_properties.py:44]: Perhaps add some logging here, to tell the user why we passed?

Example:

import logging
logging.basicConfig(level=logging.INFO)
logging.info(f"Property {XYZ} already exists")

File: sdk/tests/tutorials/properties/test_derived_properties.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    def get_instruments_with_deriv_prop(self, Figi):
+        response = self.instruments_api.list_instruments(
+                instrument_property_keys=["Instrument/Test-Demo/TestDerivedRating"],
+                filter="Identifiers[Figi] eq '{}'".format(Figi))
+
+        return response.values[0].properties[0].value.metric_value.value
+
+    # def get_instruments_with_deriv_prop(self, Figi):
+    #     response = self.instruments_api.get_instruments(identifier_type="Figi", request_body=[Figi])
+    #     return response
```
Comment (slemasne) [inline sdk/tests/tutorials/properties/test_derived_properties.py:74]: Remove this if not needed?

File: sdk/tests/tutorials/properties/test_derived_properties.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+    def test_derived_property(self):
+
+        # create instrument property edge cases
+        self.create_ratings_property("BBG000FD8G46", fitch_value = 10, moodys_value = 5)
+        self.create_ratings_property("BBG000DW76R4")
+        self.create_ratings_property("BBG000PQKVN8", moodys_value = 5)
+        self.create_ratings_property("BBG000BDWPY0", fitch_value = 10)
+
+        # create derived property using the 'Coalesce' derivation formula
+        derivation_formula = "Coalesce(Properties[Instrument/Test-Demo/MoodysRating], Properties[Instrument/Test-Demo/FitchRating],0)"
```
Comment (slemasne) [inline sdk/tests/tutorials/properties/test_derived_properties.py:85]: Nice one, well done!

File: sdk/tests/tutorials/properties/test_derived_properties.py [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+        #test case for derived property with fitch only
+        fitch_only = self.get_instruments_with_deriv_prop("BBG000BDWPY0")
+        self.assertEqual(fitch_only, 10.0)
+
+        #test case for derived property with moodys only
+        moodys_only = self.get_instruments_with_deriv_prop("BBG000PQKVN8")
+        self.assertEqual(moodys_only, 5.0)
+
+        #check why get_instruments response is empty at properties
+        # both_ratings = self.get_instruments_with_deriv_prop("BBG000FD8G46")
```
Comment (slemasne) [inline sdk/tests/tutorials/properties/test_derived_properties.py:127]: Remove if not needed.

File: sdk/tests/tutorials/properties/test_derived_properties.py [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+        try:
+            # create property definition
+            derived_prop_response = self.property_definitions_api.create_derived_property_definition(
+                derived_prop_definition_req)
+        except lusid.ApiException as e:
+            if json.loads(e.body)["name"] == "PropertyAlreadyExists":
+                pass
+
+        #self.assertGreater(len(derived_prop_response.values), 0, "Response = 0")
+        #todo add more tests for property defintion response
```
Comment (slemasne) [inline sdk/tests/tutorials/properties/test_derived_properties.py:108]: Remove if not needed

File: sdk/tests/tutorials/properties/test_derived_properties.py [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+                            models.ModelProperty(key=key,
+                                                 value=models.PropertyValue(
+                                                     metric_value=models.MetricValue(value=properties[key]))
+                                                 )]
+                    )]
+
+                response = self.instruments_api.upsert_instruments_properties(
+                    upsert_instrument_property_request=property_request)
+
+    def get_instruments_with_deriv_prop(self, Figi):
```
Comment (slemasne) [inline sdk/tests/tutorials/properties/test_derived_properties.py:65]: Maybe change the name to:
`get_instruments_with_derived_prop`
We might want to be a bit more verbose as 'deriv' can sometimes be shorthand for 'derivatives'

File: sdk/tests/tutorials/properties/test_derived_properties.py [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+        for rating in ratings:
+            property_definition = models.CreatePropertyDefinitionRequest(
+                domain="Instrument",
+                scope="Test-Demo",
+                code="{}Rating".format(rating),
+                display_name="{}Rating".format(rating),
+                data_type_id=lusid.ResourceId(
+                    scope="system", code="number"))
+
+            try:
```
Comment (slemasne) [inline sdk/tests/tutorials/properties/test_derived_properties.py:38]: Can you restructure the code, so the two properties get created once at the start?
In the current implementation, you attempt to create the properties each time you assign to a new instrument. 
This results in 6 calls to `create_property_definition` when only 2 are needed.

File: sdk/tests/tutorials/properties/test_derived_properties.py [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+                                                 value=models.PropertyValue(
+                                                     metric_value=models.MetricValue(value=properties[key]))
+                                                 )]
+                    )]
+
+                response = self.instruments_api.upsert_instruments_properties(
+                    upsert_instrument_property_request=property_request)
+
+    def get_instruments_with_deriv_prop(self, Figi):
+        response = self.instruments_api.list_instruments(
```
Comment (slemasne) [inline sdk/tests/tutorials/properties/test_derived_properties.py:66]: Why can't we use get_instrument() which is much more efficient when we know the unique IDs of the instrument we want?

Comment (cscialino) [reply sdk/tests/tutorials/properties/test_derived_properties.py:66]: See response when using get_instrument() call -- would appear that it doesn't find the properties with the given Figi, returns an empty list. Not sure if there is another way to call the properties themselves (i.e. get_property)). 

>>> response: properties[]

```
'values': {'BBG000FD8G46': {'href': 'https://swagpotato.lusid.com/api/api/instruments/LusidInstrumentId/LUID_L6GFAZUE',
                             'identifiers': {'Figi': 'BBG000FD8G46',
                                             'LusidInstrumentId': 'LUID_L6GFAZUE'},
                             'instrument_definition': None,
                             'links': [{'description': None,
                                        'href': 'https://swagpotato.lusid.com/api/api/schemas/entities/Instrument',
                                        'method': 'GET',
                                        'relation': 'EntitySchema'}],
                             'lookthrough_portfolio': None,
                             'lusid_instrument_id': 'LUID_L6GFAZUE',
                             'name': 'HISCOX LTD',
                             'properties': [],
                             'state': 'Active',
                             'version': {'as_at_date': datetime.datetime(2020, 9, 30, 9, 19, 39, 9127, tzinfo=tzutc()),
                                         'effective_from': datetime.datetime(1, 1, 1, 0, 0, tzinfo=tzutc())}}}}
```

File: sdk/tests/tutorials/properties/test_derived_properties.py (hunk 9 above)
Comment (cscialino) [reply sdk/tests/tutorials/properties/test_derived_properties.py:38]: Sorted as 2 separate functions. 

- def create_ratings_property(self, *ratings):
- def upsert_ratings_property(self, figi, fitch_value=None, moodys_value=None):

should only call `create_property_definition` by number of arguments passed.

File: sdk/tests/tutorials/properties/test_derived_properties.py (hunk 1 above)
Comment (cscialino) [reply sdk/tests/tutorials/properties/test_derived_properties.py]: Formatting completed.

**Comment by rizwansaeed** [comment]
can you fix up all the Sonar "code smells"?

File: sdk/tests/tutorials/properties/test_derived_properties.py [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+from utilities import InstrumentLoader
+from utilities import TestDataUtilities
+
+
+class DerivedPropertyTests(unittest.TestCase):
+    @classmethod
+    def setUpClass(cls):
+        # create a configured API client
+        api_client = TestDataUtilities.api_client()
+        cls.scopes_api = lusid.ScopesApi(api_client)
```
Comment (rizwansaeed) [inline sdk/tests/tutorials/properties/test_derived_properties.py:17]: this looks redundant, is it needed?

File: sdk/tests/tutorials/properties/test_derived_properties.py [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+        logging.basicConfig(level=logging.INFO)
+
+    def create_ratings_property(self, *ratings):
+
+        ratings = [*ratings]
+
+        for rating in ratings:
+            property_definition = models.CreatePropertyDefinitionRequest(
+                domain="Instrument",
+                scope="Test-Demo",
```
Comment (rizwansaeed) [inline sdk/tests/tutorials/properties/test_derived_properties.py:33]: I'm a bit of a stickler for names, can we change the scope to something descriptive like `derived` or `derivedproperties`

File: sdk/tests/tutorials/properties/test_derived_properties.py [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+
+            try:
+                # create property definition
+                self.property_definitions_api.create_property_definition(
+                    create_property_definition_request=property_definition
+                )
+            except lusid.ApiException as e:
+                if json.loads(e.body)["name"] == "PropertyAlreadyExists":
+                    logging.info(f"Property {rating} already exists")
+                    pass
```
Comment (rizwansaeed) [inline sdk/tests/tutorials/properties/test_derived_properties.py:47]: this isn't needed (unless you meant `return`)?

File: sdk/tests/tutorials/properties/test_derived_properties.py [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+
+    def upsert_ratings_property(self, figi, fitch_value=None, moodys_value=None):
+
+        properties = {
+            "Instrument/Test-Demo/FitchRating": fitch_value,
```
Comment (rizwansaeed) [inline sdk/tests/tutorials/properties/test_derived_properties.py:52]: update the scopes

File: sdk/tests/tutorials/properties/test_derived_properties.py [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+            "Instrument/Test-Demo/MoodysRating": moodys_value,
+        }
+
+        # upsert property definition
+        for key in properties:
+            if properties[key] is not None:
```
Comment (rizwansaeed) [inline sdk/tests/tutorials/properties/test_derived_properties.py:58]: this seems a necessary given  you are creating the dict above

File: sdk/tests/tutorials/properties/test_derived_properties.py [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+                # create property definition
+                self.property_definitions_api.create_property_definition(
+                    create_property_definition_request=property_definition
+                )
+            except lusid.ApiException as e:
+                if json.loads(e.body)["name"] == "PropertyAlreadyExists":
+                    logging.info(f"Property {rating} already exists")
+                    pass
+
+    def upsert_ratings_property(self, figi, fitch_value=None, moodys_value=None):
```
Comment (rizwansaeed) [inline sdk/tests/tutorials/properties/test_derived_properties.py:49]: rename to `add_rating_properties_to_instrument`

File: sdk/tests/tutorials/properties/test_derived_properties.py [hunk 17]
```diff
@@ -0,0 +1,145 @@
...
+                    upsert_instrument_property_request=property_request
+                )
+
+    def get_instruments_with_derived_prop(self, Figi):
+        response = self.instruments_api.list_instruments(
+            instrument_property_keys=["Instrument/Test-Demo/TestDerivedRating"],
+            filter=f"Identifiers[Figi] eq '{Figi}'",
+        )
+
+        return response.values[0].properties[0].value.metric_value.value
```
Comment (rizwansaeed) [inline sdk/tests/tutorials/properties/test_derived_properties.py:86]: `values` is a dict, index using the figi rather than `values[0]`

File: sdk/tests/tutorials/properties/test_derived_properties.py (hunk 10 above)
Comment (rizwansaeed) [reply sdk/tests/tutorials/properties/test_derived_properties.py:81]: this doesn't seem right, are you saying that derived properties don't come back?

File: sdk/tests/tutorials/properties/test_derived_properties.py [hunk 18]
```diff
@@ -0,0 +1,145 @@
...
+
+    def get_instruments_func(self, Figi):
```
Comment (rizwansaeed) [inline sdk/tests/tutorials/properties/test_derived_properties.py:88]: is this used (I couldn't see where)?

File: sdk/tests/tutorials/properties/test_derived_properties.py [hunk 19]
```diff
@@ -0,0 +1,145 @@
...
+        self.create_ratings_property("Fitch", "Moodys")
+
+        # create instrument property edge cases and upsert (using arbitrary numeric ratings)
+        self.upsert_ratings_property("BBG000FD8G46", fitch_value=10, moodys_value=5)
+        self.upsert_ratings_property("BBG000DW76R4")
+        self.upsert_ratings_property("BBG000PQKVN8", moodys_value=5)
+        self.upsert_ratings_property("BBG000BDWPY0", fitch_value=10)
+
+        # create derived property using the 'Coalesce' derivation formula
+        derivation_formula = "Coalesce(Properties[Instrument/Test-Demo/MoodysRating], Properties[Instrument/Test-Demo/FitchRating],0)"
```
Comment (rizwansaeed) [inline sdk/tests/tutorials/properties/test_derived_properties.py:105]: update the scope

File: sdk/tests/tutorials/properties/test_derived_properties.py [hunk 20]
```diff
@@ -0,0 +1,145 @@
...
+
+        # create derived property request
+        derived_prop_definition_req = models.CreateDerivedPropertyDefinitionRequest(
+            domain="Instrument",
+            scope="Test-Demo",
```
Comment (rizwansaeed) [inline sdk/tests/tutorials/properties/test_derived_properties.py:110]: update scope

File: sdk/tests/tutorials/properties/test_derived_properties.py [hunk 21]
```diff
@@ -0,0 +1,145 @@
...
+            code="TestDerivedRating",
```
Comment (rizwansaeed) [inline sdk/tests/tutorials/properties/test_derived_properties.py:111]: rename to `DerivedRating`

File: sdk/tests/tutorials/properties/test_derived_properties.py [hunk 22]
```diff
@@ -0,0 +1,145 @@
...
+            display_name="TestDerivedRating",
```
Comment (rizwansaeed) [inline sdk/tests/tutorials/properties/test_derived_properties.py:112]: rename to `DerivedRating`

File: sdk/tests/tutorials/properties/test_derived_properties.py [hunk 23]
```diff
@@ -0,0 +1,145 @@
...
+
+        try:
+            # create property definition
+            self.property_definitions_api.create_derived_property_definition(
+                derived_prop_definition_req
+            )
+        except lusid.ApiException as e:
+            if json.loads(e.body)["name"] == "PropertyAlreadyExists":
+                logging.info("Property 'derived' already exists")
+                pass
```
Comment (rizwansaeed) [inline sdk/tests/tutorials/properties/test_derived_properties.py:125]: remove

File: sdk/tests/tutorials/properties/test_derived_properties.py [hunk 24]
```diff
@@ -0,0 +1,145 @@
...
+            )
+
+            try:
+                # create property definition
+                self.property_definitions_api.create_property_definition(
+                    create_property_definition_request=property_definition
+                )
+            except lusid.ApiException as e:
+                if json.loads(e.body)["name"] == "PropertyAlreadyExists":
+                    logging.info(f"Property {rating} already exists")
```
Comment (slemasne) [inline sdk/tests/tutorials/properties/test_derived_properties.py:46]: Add the full property path:

Instrument/{scope}/{rating}

File: sdk/tests/tutorials/properties/test_derived_properties.py [hunk 25]
```diff
@@ -0,0 +1,145 @@
...
+            self.property_definitions_api.create_derived_property_definition(
+                derived_prop_definition_req
+            )
+        except lusid.ApiException as e:
+            if json.loads(e.body)["name"] == "PropertyAlreadyExists":
+                logging.info("Property 'derived' already exists")
+                pass
+
+        # test case for derived property with both ratings
+        both_ratings = self.get_instruments_with_derived_prop("BBG000FD8G46")
```
Comment (rizwansaeed) [inline sdk/tests/tutorials/properties/test_derived_properties.py:128]: rename `both_ratings` to `moodys_then_fitch` for clarity to the reader

**Review by rizwansaeed — CHANGES_REQUESTED** [review]

File: sdk/tests/tutorials/properties/test_derived_properties.py (hunk 10 above)
Comment (slemasne) [reply sdk/tests/tutorials/properties/test_derived_properties.py:81]: Can you share your get_instrument request?
You need to pass it the property_keys you want to see.
Example:

        looked_up_instruments = self.instruments_api.get_instruments(identifier_type="Figi",
                                                                     request_body=[figi],
                                                                     property_keys=[
                                                                         "Instrument/default/ClientInternal"
                                                                     ])

File: sdk/tests/tutorials/properties/test_derived_properties.py [hunk 26]
```diff
@@ -0,0 +1,145 @@
...
+        # create a configured API client
+        api_client = TestDataUtilities.api_client()
+        cls.scopes_api = lusid.ScopesApi(api_client)
+        cls.property_definitions_api = lusid.PropertyDefinitionsApi(api_client)
+        cls.instruments_api = lusid.InstrumentsApi(api_client)
+        # load instruments from InstrumentLoader
+        instrument_loader = InstrumentLoader(cls.instruments_api)
+        cls.instrument_ids = instrument_loader.load_instruments()
+        # setup logging configuration
+        logging.basicConfig(level=logging.INFO)
```
Comment (slemasne) [inline sdk/tests/tutorials/properties/test_derived_properties.py:24]: Add this at top under the `import` statements

**Review by slemasne — COMMENTED** [review]
3 more comments 
@cscialino

File: sdk/tests/tutorials/properties/test_derived_properties.py (hunk 10 above)
Comment (cscialino) [reply sdk/tests/tutorials/properties/test_derived_properties.py:81]: That resolved it, needed the correct property keys for the derived scope. Adjusted with get_instrument for latest push.

File: sdk/tests/tutorials/properties/test_derived_properties.py (hunk 15 above)
Comment (cscialino) [reply sdk/tests/tutorials/properties/test_derived_properties.py:58]: I believe will need it with the way the upsert propreties function is currently structured. 

It seems to otherwise loop through a "None" entry and upsert to LUSID, which is then read as value=0. 

This means in an edge case where we only have a Fitch rating but no Moody's,  it will follow the rule moodys_then_fitch from the derivation formula, and since moodys turns to 0, it assigns value 0 rather than fitch.

File: sdk/tests/tutorials/properties/test_derived_properties.py (hunk 22 above)
Comment (cscialino) [reply sdk/tests/tutorials/properties/test_derived_properties.py:112]: renamed to "Derived" in final version.
